packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -0,0 +1,252 @@
+              templateStrings.twitter,
+              source.twitter.title
+            )
+            target.twitter.title = updatedStashedTwitterTitle
+          }
+        } else {
+          target.twitter = null
+        }
+        break
+      }
+      case 'viewport': {
+        let content: string | null = null
+        const { viewport } = source
+        if (typeof viewport === 'string') {
+          content = viewport
+        } else if (viewport) {
+          content = ''
+          for (const viewportKey_ in viewPortKeys) {
+            const viewportKey = viewportKey_ as keyof Viewport
+            if (viewport[viewportKey]) {
+              if (content) content += ', '
+              content += `${viewPortKeys[viewportKey]}=${viewport[viewportKey]}`
+            }
+          }
+        }
+        target.viewport = content
+        break
+      }
+      default: {
+        // TODO: Make sure the type is correct.
+        // @ts-ignore
+        target[key] = source[key]
+        break
+      }
+    }
+  }
+}
+
+export async function resolveMetadata(metadataItems: Item[]) {
+  const resolvedMetadata = createDefaultMetadata()
+
+  let committedTitleTemplate: string | null = null
+  let committedOpenGraphTitleTemplate: string | null = null
+  let committedTwitterTitleTemplate: string | null = null
+
+  let lastLayer = 0
+  // from root layout to page metadata
+  for (let i = 0; i < metadataItems.length; i++) {
+    const item = metadataItems[i]
+    const isLayout = item.type === 'layout'
+    const isPage = item.type === 'page'
+    if (isLayout || isPage) {
+      let layerMod = await item.mod()
+
+      // Layer is a client component, we just skip it. It can't have metadata
+      // exported. Note that during our SWC transpilation, it should check if
+      // the exports are valid and give specific error messages.
+      if (
+        '$$typeof' in layerMod &&
+        (layerMod as any).$$typeof === Symbol.for('react.module.reference')
+      ) {
+        continue
+      }
+
+      if (layerMod.metadata && layerMod.generateMetadata) {
+        throw new Error(
+          `A ${item.type} is exporting both metadata and generateMetadata which is not supported. If all of the metadata you want to associate to this ${item.type} is static use the metadata export, otherwise use generateMetadata. File: ` +
+            item.path
+        )
+      }
+
+      // If we resolved all items in this layer, commit the stashed titles.
+      if (item.layer >= lastLayer) {
+        committedTitleTemplate = resolvedMetadata.title?.template || null
+        committedOpenGraphTitleTemplate =
+          resolvedMetadata.openGraph?.title?.template || null
+        const twitterTitle = resolvedMetadata.twitter?.title
+        if (
+          twitterTitle &&
+          typeof twitterTitle !== 'string' &&
+          'template' in twitterTitle
+        ) {
+          committedTwitterTitleTemplate = twitterTitle.template || null
+        }
```

Comment:
why is twitter title treated differently than the other two?

Reply:
Need to tweak bit the types of twitter, the typing didn't separate the configured twitter and resolved twitter metadata. I think that will resolve this

Reply:
Probably just an oversight on my part. If anything seems off about the RFC please update it with a comment. I tried to be thorough but it's so big :)